### PR TITLE
refactor: update interface and support asynchronous callbacks on step completion and failure

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request changes the following:

* refactor: allow steps to have asynchronous callbacks on completion and failure
* feat: allow stateful steps
* fix: code cleanup and updated example

BREAKING CHANGE: Step interface updated
- Execute now returns *Report (was (*Report, error))
- OnRollback replaced by Rollback returning *Report
- State method added to Step interface

### Related Issues

* Closes #
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

## Code coverage
Screenshot 2025-10-04 at 9 42 58 am

### CODE REVIEW

File: workflow_builder.go [hunk 1]
```diff
@@ -40,16 +38,17 @@ func (wb *workflowBuilder) Build() (Step, error) {
...
 	}
-	return NewWorkflow(wb.id, steps, WithWorkflowLogger(wb.logger), WithRollbackMode(wb.rollbackMode)), nil
-}
 
-func (wb *workflowBuilder) Steps(steps ...Builder) WorkFlowBuilder {
-	wb.mu.Lock()
-	defer wb.mu.Unlock()
+	wb.workflow.steps = steps
+	finished := wb.workflow
+	wb.workflow = newWorkflow()
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow_builder.go:44]: We may need to build multiple instances of a workflow from the same builder. So we are keeping it as it is.

File: workflow.go [hunk 2]
```diff
@@ -2,41 +2,79 @@ package automa
...
+	preparedCtx, err := wf.Prepare(ctx)
+	if err != nil {
+		return FailureReport(wf,
+			WithActionType(ActionExecute),
+			WithStartTime(start),
+			WithError(StepExecutionError.
+				Wrap(err, "workflow %q preparation failed: %v", wf.Id(), err).
+				WithProperty(StepIdProperty, wf.Id()),
+			))
+	}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:51]: Semantics is that we call handleCompletion or handleFailure based on the report from execution

[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:56]: We have added a new action type ActionPrepare, so that caller can inspect the report and determine if any cleanup needs to happen as OnCompletion wasn't called

File: workflow.go [hunk 3]
```diff
@@ -48,36 +91,89 @@ func (w *workflow) Id() string {
...
+		stepReports = append(stepReports, report)
 
+		if !report.IsSuccess() {
 			// Perform rollback for all executed steps up to the current one
-			rollbackReports := w.rollbackFrom(ctx, index)
-			if len(rollbackReports) != len(stepReports) && w.rollbackMode != RollbackModeContinueOnError {
-				return nil, errorx.IllegalState.New("mismatched rollback reports and step reports lengths, "+
-					"rollback reports: %d, step reports: %d", len(rollbackReports), len(stepReports))
-			}
+			rollbackReports := w.rollbackFrom(stepCtx, index)
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:176]: We still want that and step implementation should check before rolling back.

File: state.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+// StateFromContext retrieves the StateBag from context.
+func StateFromContext(ctx context.Context) StateBag {
+	if ctx != nil {
+		if stateBag, ok := ctx.Value(KeyState).(StateBag); ok {
+			return stateBag
+		}
+	}
+	return &SyncStateBag{}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply state.go:116]: it was intentional

File: report.go [hunk 5]
```diff
@@ -154,25 +219,45 @@ func NewReport(id string, opts ...ReportOption) *Report {
...
+
+// FailureReport constructs a failure report with options
+func FailureReport(s Step, opts ...ReportOption) *Report {
+	opts = append(opts, WithIsWorkflow(IsWorkflow(s)), WithStatus(StatusFailed), WithEndTime(time.Now()))
+	return NewReport(s.Id(), opts...)
+}
+
+// SkippedReport constructs a skipped report with options
+func SkippedReport(s Step, opts ...ReportOption) *Report {
+	opts = append(opts, WithIsWorkflow(IsWorkflow(s)), WithStatus(StatusSkipped), WithEndTime(time.Now()))
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply report.go:252]: it was intentional

File: workflow.go [hunk 6]
```diff
@@ -86,72 +182,92 @@ func (w *workflow) Execute(ctx context.Context) (*Report, error) {
...
+			)
 		}
+		stepReports = append(stepReports, report)
 	}
-	return StepSuccessReport(w.id, WithActionType(ActionRollback), WithStartTime(startTime), WithStepReports(stepReports...)), nil
+
+	return SuccessReport(w,
+		WithActionType(ActionRollback),
+		WithStartTime(startTime),
+		WithStepReports(stepReports...))
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:237]: It means the rollback itself was successful

**Review by brunodam — APPROVED** [review]
LGTM
